randomizer/bossshuffle.py

```diff
@@ -0,0 +1,150 @@
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+import random
+from doslib.rom import Rom
+from doslib.enemy import *
+from doslib.enemy import EnemyScript
+from collections import namedtuple
+from doslib.dos_utils import load_tsv
+from stream.outputstream import OutputStream
+
+NewScript = namedtuple("NewScript", ["iteration", "index", "name", "spell_chance", "ability_chance", 
+                                     "spell_1", "spell_2", "spell_3", "spell_4", "spell_5", "spell_6", 
+                                     "spell_7", "spell_8", "ability_1", "ability_2", "ability_3", "ability_4"])
+
+maxBossIndex = 0xA0
+maxScriptIndex = 0x3C
+Fiend1Offsets = [119,121,123,125]
+Fiend2Offsets = [120,122,124,126]
+Fiend1ScriptOffsets = [34,36,38,40]
+Fiend2ScriptOffsets = [35,37,39,41]
+
+class BossData:
+    def __init__(self, rom: Rom):
+        # Initialize data - the 8 boss spots that are changed, as well as the name, graphics and script blocks
+        # If we don't randomize anything, get_patches() should return the same data we load in
+        name_pointers = rom.open_bytestream(0x1DDD38, 0x280)
+        self.name_pointers = []
+        while not name_pointers.is_eos():
+            self.name_pointers.append(EnemyName(name_pointers))
+            
+        graphics_pointers = rom.open_bytestream(0x2227D8, 0x780)
+        self.graphics_pointers = []
+        while not graphics_pointers.is_eos():
+            self.graphics_pointers.append(EnemyGraphics(graphics_pointers))
+            
+        attack_animations = rom.open_bytestream(0x223540, 0xA0)
+        self.attack_animations = []
+        while not attack_animations.is_eos():
+            self.attack_animations.append(attack_animations.get_u8())
+            
+        enemy_script_stream = rom.open_bytestream(0x22F17C, 0x450)
+        self.scripts = []
+        while not enemy_script_stream.is_eos():
+            self.scripts.append(EnemyScript(enemy_script_stream))
+            
+        self.bossData = {}
+        for script in load_tsv("data/BossScriptData.tsv"):
+            entry = NewScript(*script)
+            if entry.name not in self.bossData:
+                self.bossData[entry.name] = [None,None,None]
+            self.bossData[entry.name][entry.iteration] = entry
+        self.bossList = list(self.bossData.keys())
+        
+        #Finally, load in the script data from the ROM - we can read in the tsv if/when randomize gets called
+    
+    def randomize_bosses(self, rng: random.Random):
+        bossChoices = rng.sample(self.bossList,4)
+        rng.shuffle(bossChoices)
+        newFiend1s = []
+        newFiend1s.append(self.bossData[bossChoices[0]][0])
```

Comment:
Consider explaining (or making constants for) the second index. I _think_ 0 is easy, 1 is normal, and 2 is hard (revisited) versions, but I won't remember that in a week ;)